js/web/lib/wasm/jsep/backend-webgpu.ts

```diff
@@ -0,0 +1,270 @@
+          createTemporaryOutput(programInfo.outputs[i].dataType, programInfo.outputs[i].dims) :
+          createKernelOutput(validatedOutputIndices[i], programInfo.outputs[i].dataType, programInfo.outputs[i].dims);
+      const gpuData = this.gpuDataManager.get(tensorView.data);
+      if (!gpuData) {
+        throw new Error(`no GPU data for output: ${tensorView.data}`);
+      }
+      if (isTemporary) {
+        this.temporaryData.push(gpuData);
+      }
+      if (isPersistent) {
+        let persistentData = this.kernelPersistentData.get(this.currentKernelId!);
+        if (!persistentData) {
+          persistentData = [];
+          this.kernelPersistentData.set(this.currentKernelId!, persistentData);
+        }
+        persistentData.push(gpuData);
+      }
+      outputTensorViews.push(tensorView);
+      outputDatas.push(gpuData);
+    }
+
+    if (!artifact) {
+      artifact = this.programManager.build(programInfo);
+      this.programManager.setArtifact(key, artifact);
+    }
+
+    this.programManager.run(artifact, inputDatas, outputDatas, artifact.programInfo.dispatchGroup(inputs));
+
+    return outputTensorViews;
+  }
+
+  upload(gpuDataId: number, data: Uint8Array): void {
+    this.gpuDataManager.upload(gpuDataId, data);
+  }
+
+  memcpy(src: number, dst: number): void {
+    this.gpuDataManager.memcpy(src, dst);
+  }
+
+  async download(gpuDataId: number, data: Uint8Array): Promise<void> {
+    const arrayBuffer = await this.gpuDataManager.download(gpuDataId);
+    data.set(new Uint8Array(arrayBuffer));
+  }
+
+  alloc(size: number): number {
+    return this.gpuDataManager.create(size).id;
+  }
+
+  free(ptr: number): number {
+    return this.gpuDataManager.release(ptr);
+  }
+
+  createKernel(name: string, kernelId: number, attribute: unknown): void {
+    const op = WEBGPU_OP_RESOLVE_RULES.get(name);
+    if (!op) {
+      throw new Error(`kernel not implemented: ${name}`);
+    }
+
+    this.kernels.set(kernelId, [name, op[0], [op[1], attribute]]);
+  }
+
+  releaseKernel(kernelId: number): void {
+    const persistentData = this.kernelPersistentData.get(kernelId);
+    if (persistentData) {
+      for (const data of persistentData) {
+        this.gpuDataManager.release(data.id);
+      }
+      this.kernelPersistentData.delete(kernelId);
+    }
+    this.kernels.delete(kernelId);
+  }
+
+  computeKernel(kernelId: number, context: ComputeContext): number {
+    const kernel = this.kernels.get(kernelId);
+    if (!kernel) {
+      throw new Error(`kernel not created: ${kernelId}`);
+    }
+    const [name, kernelEntry, attributes] = kernel;
+    if (this.currentKernelId !== null) {
+      throw new Error(`kernel "${name}" is not allowed to be called recursively`);
+    }
+    this.currentKernelId = kernelId;
+
+    // parse attributes if necessary
+    if (attributes[0]) {
+      attributes[1] = attributes[0](attributes[1]);
+      attributes[0] = undefined;
+    }
```

Comment:
each kernel implementation may or may not include a call-once callback for pre-processing the attributes. if it is present, the pre-processing will be called once so that the parsed attributes will be used as parameters. either way, we don't need to parse attributes on each call to compute; we simply pass it to the compute function as parameter.